### PR TITLE
Experiment

### PR DESCRIPTION
**Pull request by katabana**

### CODE REVIEW

File: experiment.py [hunk 1]
```diff
@@ -16,252 +21,127 @@
...
+
+        if TWO_DIMENSIONAL:
+            dir_path = os.path.join(generated_dir_path)
+            try:
+                helpers.make_dirs(dir_path, exist_ok=True)
+            except OSError:
+                if not os.path.isdir(dir_path):
+                    raise
+
+            f = open(os.path.join(dir_path, name + ".in"), "w")
```
Comment (pbatko) [inline experiment.py]: what about using `with open(..) as f:` ?

File: experiment.py [hunk 2]
```diff
@@ -16,252 +21,127 @@
...
+
+    def run(self):
+        for experiment_command in self.__commands:
+            {
+                'save': lambda x: self.save_data(x),
+                'compute_winners': lambda x: self.compute_winners(*x)
+            }[experiment_command[0]](experiment_command[1])
+
+    # save data
+    def save_data(self, name):
```
Comment (pbatko) [inline experiment.py]: `name` - is it just file name? or a whole path?
have you considered using some lib for handling paths? e.g. pathlib(2?)

File: experiment.py (hunk 1 above)
Comment (katabana) [reply experiment.py]: Right, I will remember about it in the future but for now I'm trying to delete most of operations on files. Logging and storing temporary results will be added later. My focus now is on calculating it without reading and saving to files.

File: experiment.py (hunk 2 above)
Comment (katabana) [reply experiment.py]: I remember about it, I know it's messy but I haven't thought the structure through yet.

File: experiment_config.py [hunk 3]
```diff
@@ -87,23 +92,23 @@ def get_generated_dir_path(self):
     def set_candidates(self, list_of_candidates):
         self.__candidates = list_of_candidates
 
-    def add_candidates_function(self, generating_function):
-        self.__candidates_generating += [generating_function]
-
     def add_candidates(self, list_of_candidates):
-        self.__candidates += list_of_candidates
+        if inspect.isfunction(list_of_candidates):
```
Comment (pbatko) [inline experiment_config.py]: I think the previous approach with a separate method for accepting functions was better (less hacky)

**Pull request merged**